File: tests/public_generic_native_adapter_v1/authenticated_handoff/same_subject_lifecycle.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
size_t auth_live(void);
size_t auth_calls(void);
int main(void) {
    assert(printf("[") == 1);
    for (size_t cycle = 0; cycle < 2; ++cycle) {
        spx_pg_provider_v1 *provider = NULL;
        assert(spx_pg_provider_open_v1(descriptor, sizeof(descriptor), binding, sizeof(binding), &provider) == 0);
        spx_pg_provider_v1 *stale_provider = provider;
        spx_pg_result_v1 *result = NULL;
        assert(spx_pg_call_v1(provider, NULL, &result) == 8 && result == NULL);
        assert(auth_calls() == cycle);
        uint64_t generation = 0;
        assert(spx_pg_authenticated_generation_v1(provider, &generation) == 0 && generation != 0);
        spx_pg_value_v1 *input = NULL;
        assert(spx_pg_authenticated_input_prepare_v1(provider, generation, 0, cleanup, sizeof(cleanup),
            canonical, sizeof(canonical), &input) == 0);
        assert(spx_pg_provider_close_v1(&provider) == 7 && provider != NULL);
        /* Native null provider is 13, unlike the Core-Wasm invalid-handle 8. */
        assert(spx_pg_call_v1(NULL, input, &result) == 13 && result == NULL);
        assert(auth_calls() == cycle);
        assert(spx_pg_call_v1(provider, input, &result) == EXPECT_CALL_STATUS);
        assert(auth_calls() == cycle + 1);
        spx_pg_result_v1 *duplicate = NULL;
        assert(spx_pg_call_v1(provider, input, &duplicate) == 8 && duplicate == NULL);
        assert(spx_pg_call_v1(provider, NULL, &duplicate) == 8 && duplicate == NULL);
        uint8_t out[31], retry[31]; size_t required = 0;
        if (EXPECT_CALL_STATUS == 0) {
            assert(result != NULL);
            assert(spx_pg_provider_close_v1(&provider) == 7 && provider != NULL);
            assert(spx_pg_result_export_v1(result, NULL, 0, &required) == 12 && required == sizeof(out));
            memset(out, 0xa5, sizeof(out));
            assert(spx_pg_result_export_v1(result, out, sizeof(out) - 1, &required) == 12 && required == sizeof(out));
            for (size_t i = 0; i < sizeof(out); ++i) assert(out[i] == 0xa5);
            assert(spx_pg_result_export_v1(result, out, sizeof(out), &required) == 0 && required == sizeof(out));
            assert(spx_pg_result_export_v1(result, retry, sizeof(retry), &required) == 0 && memcmp(out, retry, sizeof(out)) == 0);
            spx_pg_result_v1 *stale = result;
            assert(spx_pg_result_release_v1(&result) == 0 && result == NULL);
            size_t stale_required = 99;
            assert(spx_pg_result_export_v1(stale, retry, sizeof(retry), &stale_required) == 8 && stale_required == 0);
            assert(spx_pg_result_release_v1(&stale) == 8 && stale == NULL);
        } else assert(result == NULL);
        assert(spx_pg_value_release_v1(&input) == 8 && input == NULL);
        assert(auth_calls() == cycle + 1 && spx_pg_test_live_handles_v1(provider) == 0);
        assert(spx_pg_provider_close_v1(&provider) == 0 && provider == NULL);
        assert(spx_pg_provider_close_v1(&stale_provider) == 8 && stale_provider == NULL);
        assert(auth_live() == 0);
        if (cycle != 0) assert(printf(",") == 1);
        assert(printf("[") == 1);
        for (size_t i = 0; i < required; ++i) {
            if (i != 0) assert(printf(",") == 1);
            assert(printf("%u", (unsigned)out[i]) > 0);
        }
        assert(printf("]") == 1);
    }
    assert(printf("]") == 1);
    return 0;
}

File: tests/public_generic_native_adapter_v1/authenticated_handoff/same_subject_lifecycle.rs
//! Physical API lifecycle agreement, not injected-failure/heap parity.
//! Native frees owned allocations; Core-Wasm releases opaque slots in retained
//! linear memory. Failed-call transfer differs and is asserted per target.
use semaprax::public_generic_abi::{
    carrier::{
        frame::{parse_bounded, CarrierFrameBinding, CarrierLeaf, LeafKind},
        trace::Direction,
    },
    compiler_endpoint::derive_admitted_public_generic_endpoint_v1,
    native::{authenticated, template},
};
use std::{env, fs, path::Path, process::Command};

fn run(command: &mut Command) -> Vec<Vec<u8>> {
    let output = command
        .output()
        .expect("focused lifecycle gate requires clang and Node");
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    serde_json::from_slice(&output.stdout).unwrap()
}

#[test]
fn checked_identity_native_and_core_wasm_refuse_absent_handles_and_retry_export() {
    let root = env::temp_dir().join(format!(
        "semaprax-lifecycle-parity-{}-{}",
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos()
    ));
    fs::create_dir(&root).unwrap();
    eprintln!("physical lifecycle evidence: {}", root.display());
    for guard in [true, false] {
        let directory = root.join(guard.to_string());
        fs::create_dir(&directory).unwrap();
        let result = prepare(&directory, guard);
        compare(&directory, guard, &result);
    }
}

fn prepare(directory: &Path, guard: bool) -> CarrierFrameBinding {
    let source = super::SOURCE.replace(
        "requires true",
        if guard {
            "requires true"
        } else {
            "requires false"
        },
    );
    let parsed = semaprax::check(&source, Path::new("same-subject.spx")).unwrap();
    let revision = semaprax::format::canonical(&parsed);
    let program = semaprax::hir::resolve(&parsed).unwrap();
    let endpoint =
        derive_admitted_public_generic_endpoint_v1(&program, &revision, "auth.identity").unwrap();
    let native = authenticated::render_authenticated_identity_provider(
        &program,
        &revision,
        endpoint.descriptor(),
    )
    .unwrap();
    let wasm = semaprax::wasm::emit_public_generic_wasm_provider_v1(&program, &endpoint).unwrap();
    wasm.verify().unwrap();
    assert_eq!(native.descriptor_bytes(), wasm.descriptor_bytes());
    let input =
        CarrierFrameBinding::from_verified_descriptor(endpoint.descriptor(), Direction::Input);
    let result =
        CarrierFrameBinding::from_verified_descriptor(endpoint.descriptor(), Direction::Result);
    let frame = input
        .frame_with_leaves(
            input
                .leaf_paths()
                .iter()
                .zip(super::PAYLOADS)
                .map(|(path, payload)| CarrierLeaf::new(path, LeafKind::Bytes, payload.to_vec()))
                .collect(),
        )
        .encode();
    for (name, bytes) in [
        ("provider.wasm", wasm.wasm()),
        ("descriptor.bin", wasm.descriptor_bytes()),
        ("binding.bin", &wasm.binding_bytes()),
        ("input.bin", &frame),
    ] {
        fs::write(directory.join(name), bytes).unwrap();
    }
    fs::write(
        directory.join("probe.mjs"),
        include_str!("same_subject_lifecycle.mjs"),
    )
    .unwrap();
    let expected = if guard { "0" } else { "11" };
    let mut constants = String::new();
    for (name, bytes) in [
        ("descriptor", native.descriptor_bytes()),
        ("binding", &native.binding().encode()),
        (
            "cleanup",
            endpoint.descriptor().settlement().digest().as_bytes(),
        ),
        ("canonical", &frame),
    ] {
        constants.push_str(&super::array(name, bytes));
    }
    let provider = format!("{}\nstatic size_t endpoint_calls;\n#define SPX_PG_OBSERVE_ENDPOINT() (++endpoint_calls)\n{}\n#undef malloc\n#undef free\nsize_t auth_live(void) {{ return fixture_live; }}\nsize_t auth_calls(void) {{ return endpoint_calls; }}\n",
            include_str!("../allocations.c"), native.source());
    let driver = format!(
        "{}\n{}\n{constants}\n#define EXPECT_CALL_STATUS {expected}\n{}",
        template::HEADER_V1,
        authenticated::HEADER,
        include_str!("same_subject_lifecycle.c")
    );
    fs::write(directory.join("provider.c"), provider).unwrap();
    fs::write(directory.join("driver.c"), driver).unwrap();
    result
}

fn compare(directory: &Path, guard: bool, result: &CarrierFrameBinding) {
    let expected = if guard { "0" } else { "11" };
    let wasm_outputs = run(Command::new("node")
        .current_dir(directory)
        .args(["probe.mjs", expected]));
    assert_eq!(wasm_outputs.len(), 2);
    let wasm_leaves: Vec<Vec<Vec<u8>>> = wasm_outputs
        .iter()
        .map(|bytes| {
            if !guard {
                assert!(bytes.is_empty());
                return Vec::new();
            }
            let frame = parse_bounded(bytes).unwrap();
            result.validate_frame(&frame).unwrap();
            frame
                .leaves()
                .iter()
                .map(|leaf| leaf.payload().to_vec())
                .collect()
        })
        .collect();
    for opt in ["-O0", "-O2"] {
        let executable = directory.join(format!("probe{opt}{}", env::consts::EXE_SUFFIX));
        let output = Command::new("clang")
            .args(["-std=c11", opt, "-Wall", "-Wextra", "-Werror"])
            .arg(directory.join("provider.c"))
            .arg(directory.join("driver.c"))
            .arg("-o")
            .arg(&executable)
            .output()
            .unwrap();
        assert!(
            output.status.success(),
            "{}",
            String::from_utf8_lossy(&output.stderr)
        );
        let native_outputs = run(&mut Command::new(executable));
        assert_eq!(native_outputs.len(), 2);
        let native_leaves: Vec<_> = native_outputs
            .iter()
            .map(|bytes| {
                if guard {
                    super::decode_native(bytes)
                } else {
                    assert!(bytes.is_empty());
                    Vec::new()
                }
            })
            .collect();
        assert_eq!(native_leaves, wasm_leaves);
        if guard {
            assert_eq!(
                native_leaves,
                vec![super::PAYLOADS.map(<[u8]>::to_vec).to_vec(); 2]
            );
        }
        eprintln!("requires={guard} native{opt}/Core-Wasm: two cycles, status{expected}, absent/stale refusals, target-local cleanup; success export12->0 retry byte-exact");
    }
}
